### PR TITLE
LUCENE-9031: test unified highlighter on intervals queries

### PR DESCRIPTION
**Pull request by mkhludnev**
# Description

Fix UnsupportedOpExc on highlighting Interval queries. 

# Solution

Caching and returning Interval queries. 

# Tests

Cruel harm to TestUnifiedHighlighting

### CODE REVIEW

File: lucene/queries/src/java/org/apache/lucene/queries/intervals/IntervalMatches.java [hunk 1]
```diff
@@ -74,7 +75,7 @@ public MatchesIterator getSubMatches() throws IOException {
 
       @Override
       public Query getQuery() {
-        throw new UnsupportedOperationException();
+        return query==null ? query = new IntervalQuery(field, intervalsSource) : query;
```
[the remark this answers is not in the document]
Comment (mkhludnev) [reply lucene/queries/src/java/org/apache/lucene/queries/intervals/IntervalMatches.java]: well .. It's a static method, hence, there's no query instance, even IntervalSources which invoke this method.. So, don't know where to get  ```source.getQuery()```, but I like the idea.

File: lucene/queries/src/java/org/apache/lucene/queries/intervals/TermIntervalsSource.java [hunk 2]
```diff
@@ -200,7 +207,7 @@ public MatchesIterator getSubMatches() {
 
       @Override
       public Query getQuery() {
-        throw new UnsupportedOperationException();
+        return query;
```
[the remark this answers is not in the document]
Comment (mkhludnev) [reply lucene/queries/src/java/org/apache/lucene/queries/intervals/TermIntervalsSource.java]: I don't follow... Where I can find how ```SpanQuery returns a TermQuery here``` ? I'll try to do something like this, but how it's worth to do it eagerly or lazily?

[the remark this answers is not in the document]
Comment (mkhludnev) [reply lucene/queries/src/java/org/apache/lucene/queries/intervals/TermIntervalsSource.java]: Couldn't make it better, upload almost as is.

File: lucene/queries/src/java/org/apache/lucene/queries/intervals/CachingMatchesIterator.java [hunk 3]
```diff
@@ -27,8 +27,9 @@
 class CachingMatchesIterator extends FilterMatchesIterator {
 
   private boolean positioned = false;
-  private int[] posAndOffsets = new int[16];
-  private int count = 0;
+  private int[] posAndOffsets = new int[4*4];
+  private Query[] mathchingQueries = new Query[4];
```
[the remark this answers is not in the document]
Comment (mkhludnev) [reply lucene/queries/src/java/org/apache/lucene/queries/intervals/CachingMatchesIterator.java]: fixed. uploaded

File: lucene/queries/src/java/org/apache/lucene/queries/intervals/IntervalMatches.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (romseygeek) [reply lucene/queries/src/java/org/apache/lucene/queries/intervals/IntervalMatches.java]: I mean literally replace `throw new UnsupporrtedOperationException` with `return source.getQuery()` - just delegate, like we're doing with submatches four lines above.

File: lucene/queries/src/java/org/apache/lucene/queries/intervals/TermIntervalsSource.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (romseygeek) [reply lucene/queries/src/java/org/apache/lucene/queries/intervals/TermIntervalsSource.java]: Have a look at how `SpanWeight` implements this.

**Pull request closed without merging**